File: Java/HouseRobber.java
class Solution {
    private int[] cache;

    public int solve(int idx, int[] nums) {
        if(idx >= nums.length) return 0;
        if(cache[idx] >= 0) return cache[idx];
        int max = 0;
        max = Math.max(nums[idx] + solve(idx + 2, nums), solve(idx + 1, nums));
        cache[idx] = max;
        return max;
    }

    public int rob(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        cache = new int[nums.length];
        for(int i = 0; i < cache.length; i++) {
            cache[i] = -1;
        }
        return solve(0, nums);    
    }
	
	public int rob2(int[] nums) {
        if(nums == null || nums.length == 0) return 0;
        int n = nums.length;
        if(n == 1) return nums[0];
        int[] d = new int[n];
        d[0] = nums[0];
        d[1] = Math.max(nums[0], nums[1]);
        int res = Math.max(d[0], d[1]);

        for(int i = 2; i < n; i++) {
            d[i] = Math.max(d[i - 2] + nums[i], d[i - 1]);
            res = Math.max(res, d[i]);
        }

        return res;
    }
}

File: Go/house_robber.go
	func rob(nums []int) int {
    var res int
    cache := make([]int, len(nums))
    for i,_ := range cache {
        cache[i] = -1
    }
    if len(nums) == 0 {return 0}

    var solve func(idx int) int
    solve = func(idx int) int {
        if idx >= len(nums) {return 0}
        if cache[idx] >= 0 {return cache[idx]}

        ret := max(nums[idx] + solve(idx + 2), solve(idx + 1))
        cache[idx] = ret
        return ret
    }
    res = solve(0)
    return res
}

func max(a, b int) int {
    if a > b {return a}
    return b
}